File: Backend/src/main/java/com/meltic/gmao/service/TelemetriaService.java
package com.meltic.gmao.service;

import com.meltic.gmao.model.nosql.Telemetria;
import com.meltic.gmao.repository.nosql.TelemetriaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.aggregation.*;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;

@Service
public class TelemetriaService {

    private static final int MAX_HISTORIC_POINTS = 2000;

    @Autowired
    private TelemetriaRepository telemetriaRepository;

    @Autowired
    private MongoTemplate mongoTemplate;

    public Telemetria guardar(Telemetria telemetria) {
        if (telemetria.getTimestamp() == null) {
            telemetria.setTimestamp(Instant.now());
        }
        return telemetriaRepository.save(telemetria);
    }

    /**
     * Carga inicial: devuelve los últimos 3600 registros ordenados ASC.
     * (preload del historian live — equivalente a abrir una pantalla SCADA)
     */
    public List<Telemetria> obtenerPorMaquina(Long maquinaId) {
        return telemetriaRepository.findTop3600ByMaquinaIdOrderByTimestampAsc(maquinaId);
    }

    /**
     * Consulta incremental SCADA: solo los registros más nuevos que `since`.
     * (report-by-exception / OPC DA subscription update)
     */
    public List<Telemetria> obtenerDesde(Long maquinaId, Instant since) {
        return telemetriaRepository.findByMaquinaIdAndTimestampAfterOrderByTimestampAsc(maquinaId, since);
    }

    /**
     * Consulta histórica multi-escala:
     * Devuelve hasta MAX_HISTORIC_POINTS puntos representativos de cualquier rango temporal.
     *
     * Algoritmo:
     *  1. Si el rango contiene ≤ MAX_HISTORIC_POINTS registros → devuelve todos (exactos).
     *  2. Si hay más → usa MongoDB $sample para muestreo estadístico uniforme + $sort ASC.
     *     Esto es equivalente al "data compression" de PI System o al "decimation" de Wonderware.
     *
     * Resultado: siempre ≤ 2000 puntos, renderizables en cualquier gráfica,
     * independientemente del rango temporal (1 día, 6 meses, 10 años).
     */
    public List<Telemetria> obtenerHistorico(Long maquinaId, Instant desde, Instant hasta) {
        if (desde == null || hasta == null) return List.of();
        MatchOperation match = Aggregation.match(
            Criteria.where("maquinaId").is(maquinaId)
                    .and("timestamp").gte(desde).lte(hasta)
        );
        SortOperation sort = Aggregation.sort(Sort.by(Sort.Direction.ASC, "timestamp"));

        // Contar primero para saber si necesitamos muestrear
        if (maquinaId == null || desde == null || hasta == null) return List.of();
        long total = mongoTemplate.count(
            org.springframework.data.mongodb.core.query.Query.query(
                Criteria.where("maquinaId").is(maquinaId)
                        .and("timestamp").gte(desde).lte(hasta)
            ),
            Telemetria.class
        );

        Aggregation agg;
        if (total <= MAX_HISTORIC_POINTS) {
            // Rango pequeño: devolver todos ordenados
            agg = Aggregation.newAggregation(match, sort);
        } else {
            // Rango grande: muestreo aleatorio representativo + ordenar por tiempo
            // $sample elige N documentos al azar del conjunto filtrado (eficiente en MongoDB 4+)
            SampleOperation sample = Aggregation.sample(MAX_HISTORIC_POINTS);
            agg = Aggregation.newAggregation(match, sample, sort);
        }

        AggregationResults<Telemetria> results = mongoTemplate.aggregate(agg, "telemetria", Telemetria.class);
        return results.getMappedResults();
    }
}

File: Frontend/meltic_gmao_app/lib/services/telemetria_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_config.dart';
import 'app_session.dart';
import '../models/telemetria.dart';

class TelemetriaService {
  /// Carga inicial: devuelve los últimos 3600 registros (buffer preload).
  /// Equivalente al historico inicial que un SCADA carga al abrir una pantalla.
  Future<List<Telemetria>> fetchPorMaquina(int maquinaId) async {
    final response = await http.get(
      Uri.parse('${ApiConfig.baseUrl}/api/plc/maquina/$maquinaId'),
      headers: AppSession.instance.authHeaders,
    );
    if (response.statusCode == 200) {
      final List<dynamic> body = json.decode(response.body);
      return body.map((item) => Telemetria.fromJson(item)).toList();
    }
    throw Exception('Error carga inicial: ${response.statusCode}');
  }

  /// Polling incremental SCADA: devuelve SOLO los registros más nuevos que [since].
  /// En un historian real esto equivale al "subscription update" o "report-by-exception".
  /// El servidor filtra en BD — el cliente recibe 0–N puntos, nunca miles.
  Future<List<Telemetria>> fetchDesde(int maquinaId, DateTime since) async {
    final epochMs = since.toUtc().millisecondsSinceEpoch;
    final response = await http.get(
      Uri.parse('${ApiConfig.baseUrl}/api/plc/maquina/$maquinaId?since=$epochMs'),
      headers: AppSession.instance.authHeaders,
    );
    if (response.statusCode == 200) {
      final List<dynamic> body = json.decode(response.body);
      return body.map((item) => Telemetria.fromJson(item)).toList();
    }
    throw Exception('Error polling incremental: ${response.statusCode}');
  }

  /// Consulta histórica multi-escala: cualquier rango temporal → ≤ 2000 puntos.
  /// El servidor aplica downsampling estadístico (MongoDB \$sample).
  /// Usar cuando la ventana supera 1 día (semanas, meses, años).
  Future<List<Telemetria>> fetchHistorico(int maquinaId, DateTime desde, DateTime hasta) async {
    final desdeMs = desde.toUtc().millisecondsSinceEpoch;
    final hastaMs = hasta.toUtc().millisecondsSinceEpoch;
    final response = await http.get(
      Uri.parse('${ApiConfig.baseUrl}/api/plc/maquina/$maquinaId/historico?desde=$desdeMs&hasta=$hastaMs'),
      headers: AppSession.instance.authHeaders,
    );
    if (response.statusCode == 200) {
      final List<dynamic> body = json.decode(response.body);
      return body.map((item) => Telemetria.fromJson(item)).toList();
    }
    throw Exception('Error histórico: ${response.statusCode}');
  }
}
